Explain planning of text indexes (#1270)
neo4j/neo4j-documentation: cypher/cypher-docs/src/docs/dev/query-tuning-indexes.adoc

|prefix search| `STARTS WITH`
|suffix search| `ENDS WITH`
|substring search| `CONTAINS`
|===

=== TEXT indexes

`TEXT` indexes only work for predicates operating on strings.
That means that `TEXT` indexes are only used when it is known that the predicate evaluates to `null` for all non-string values.

Predicates that only operate on strings are always solvable by a `TEXT` index:

* `STARTS WITH`
* `ENDS WITH`
* `CONTAINS`

However, other predicates are only used when it is known that the property is compared to a string:

* `n.prop = "string"`
* `n.prop IN ["a", "b", "c"]`
* `n.prop > "string"`

This means that a `TEXT` index is not able to solve e.g. `a.prop = b.prop`.

In summary, `TEXT` indexes support the following predicates:

[options="header"]
|===
|Predicate | Syntax
|equality check| `n.prop = "string"`
|list membership check | `n.prop IN ["a", "b", "c"]`
|range search| `n.prop > "string"`
|prefix search| `STARTS WITH`
|suffix search| `ENDS WITH`
|substring search| `CONTAINS`
|===

== Index preference

When multiple indexes are available and able to solve a predicate, there is an order defined that decides which index to use.
It is defined as such:

> * `TEXT` indexes are preferred over `BTREE` indexes.

Reviewer: TEXT should only be preferred over BTREE for CONTAINS and ENDS WITH. In all other cases BTREE should be more efficient. This should be reflected by `IndexCapability#getCostMultiplier( IndexQueryType... queryTypes )`. Is that not utilised by the planner? Ping @hedengran

Please refer to <<administration-indexes-search-performance>> for instructions on how to create and maintain the indexes themselves.

== Index types and predicate compatibility

> There are different types of indexes available in Neo4j.

Reviewer 2: ```suggestion
There are different types of indexes available in Neo4j but they are not all compatible with the same property predicates.
```

> Not all indexes are compatible with the same property predicates.

Reviewer 2: ```suggestion

```

> In the most common case, indexes are used for `MATCH` and `OPTIONAL MATCH` clauses that combine a label predicate with a property predicate.

Reviewer 2: ```suggestion
Indexes are commonly used for `MATCH` and `OPTIONAL MATCH` clauses that combine a label predicate with a property predicate.
```

> Therefore, it is important to know the kind of predicates that can be solved by the different indexes.

Reviewer 2: ```suggestion
Therefore, it is important to know what kind of predicates that can be solved by the different indexes.
```

=== BTREE indexes

`BTREE` indexes support all types of predicates:

[options="header"]
|===
|Predicate | Syntax
|equality check| `n.prop = value`
|list membership check| `n.prop IN list`
|existence check| `n.prop IS NOT NULL`
|range search| `n.prop > value`
|prefix search| `STARTS WITH`
|suffix search| `ENDS WITH`
|substring search| `CONTAINS`
|===

=== TEXT indexes

`TEXT` indexes only work for predicates operating on strings.

> That means that `TEXT` indexes will only be used when it is known that the predicate will evaluate to `null` for all non-string values.

Reviewer 2: ```suggestion
That means that `TEXT` indexes are only used when it is known that the predicate evaluates to `null` for all non-string values.
```

Predicates that only operate on strings are always solvable by a `TEXT` index:

* `STARTS WITH`
* `ENDS WITH`
* `CONTAINS`

> Other predicates, however, will only be used when it is known that the property is compared to a string:

Reviewer 2: ```suggestion
However, other predicates are only used when it is known that the property is compared to a string:
```

* `n.prop = "string"`
* `n.prop IN ["a", "b", "c"]`
* `n.prop > "string"`

> That means that a `TEXT` index will not be able to solve e.g. `a.prop = b.prop`.

Reviewer 2: ```suggestion
This means that a `TEXT` index is not able to solve e.g. `a.prop = b.prop`.
```

In summary, `TEXT` indexes support the following predicates:

[options="header"]
|===
|Predicate | Syntax
|equality check| `n.prop = "string"`
|list membership check | `n.prop IN ["a", "b", "c"]`
|range search| `n.prop > "string"`
|prefix search| `STARTS WITH`
|suffix search| `ENDS WITH`
|substring search| `CONTAINS`
|===

== Index preference

> When multiple indexes are available and able to solve a predicate there is an ordering defined that decide on which index to use.

Reviewer 2: ```suggestion
When multiple indexes are available and able to solve a predicate, there is an order defined that decides which index to use.
```

> It is defined such that:

Reviewer 2: ```suggestion
It is defined as such:
```

> * `TEXT` indexes will be preferred over `BTREE` indexes.

Reviewer 2: ```suggestion
* `TEXT` indexes are preferred over `BTREE` indexes.
```

Please refer to <<administration-indexes-search-performance>> for instructions on how to create and maintain the indexes themselves.

== Index types and predicate compatibility

There are different types of indexes available in Neo4j.
Not all indexes are compatible with the same property predicates.

In the most common case, indexes are used for `MATCH` and `OPTIONAL MATCH` clauses that combine a label predicate with a property predicate.
Therefore, it is important to know the kind of predicates that can be solved by the different indexes.

=== BTREE indexes

> `BTREE` indexes support all types of predicates:

Reviewer 3: Should we mention that BTREE allows composite indexes while text doesn't?

Reviewer 4: I have been thinking about it, too.
It is a question if this is the place. We don't allow creating a composite text index, so a user will definitely not be wondering "why did the planner not pick up my composite text index?"

Reviewer 3: true
